QuantConnect.IBAutomater/IBAutomater.cs

```diff
@@ -97,115 +133,225 @@ public IBAutomater(string ibDirectory, string ibVersion, string userName, string
+                process.BeginErrorReadLine();
+                process.BeginOutputReadLine();
 
-            if (waitForExit)
-            {
-                process.WaitForExit();
+                if (waitForExit)
+                {
+                    process.WaitForExit();
+                }
+                else
+                {
+                    // wait for completion of IBGateway login and configuration
+                    var message = _ibAutomaterInitializeEvent.WaitOne(TimeSpan.FromSeconds(60))
+                        ? "IB Automater initialized."
+                        : "IB Automater initialization timeout.";
+                    OutputDataReceived?.Invoke(this, new OutputDataReceivedEventArgs(message));
+
+                    if (_lastStartResult.HasError)
+                    {
+                        message = $"IBAutomater error - Code: {_lastStartResult.ErrorCode} Message: {_lastStartResult.ErrorMessage}";
+                        OutputDataReceived?.Invoke(this, new OutputDataReceivedEventArgs(message));
+
+                        return _lastStartResult;
+                    }
+                }
             }
+
+            return StartResult.Success;
         }
 
         /// <summary>
-        /// Stops the IB Automater
+        /// Stops the IB Gateway
         /// </summary>
         /// <remarks>The IB Gateway application will be terminated</remarks>
         public void Stop()
         {
-            if (IsWindows)
+            lock (_locker)
             {
-                foreach (var process in Process.GetProcesses())
+                if (!IsRunning())
                 {
-                    try
+                    return;
+                }
+
+                if (IsWindows)
+                {
+                    foreach (var process in Process.GetProcesses())
                     {
-                        if (process.MainWindowTitle.ToLower().Contains("ib gateway"))
+                        try
+                        {
+                            if (process.MainWindowTitle.ToLower().Contains("ib gateway"))
+                            {
+                                process.Kill();
+                            }
+                        }
+                        catch (Exception)
                         {
-                            process.Kill();
-                            Thread.Sleep(2500);
+                            // ignored
                         }
                     }
+                }
+                else
+                {
+                    try
+                    {
+                        Process.Start("pkill", "xvfb-run");
+                        Process.Start("pkill", "java");
+                        Process.Start("pkill", "Xvfb");
+                    }
                     catch (Exception)
                     {
                         // ignored
                     }
                 }
+
+                _process = null;
```

Comment:
Should we try to kill `_process` too?

Reply:
`_process` is the parent process (sh/bat script) for the processes killed in the previous lines of code and should be terminated automatically.
I did a quick run with `_process?.Kill();` added and this was the error:
```
System.InvalidOperationException: Cannot process request because the process (14300) has exited.
```